src/hotspot/share/opto/superword.cpp

```diff
@@ -491,245 +487,114 @@ bool SuperWord::SLP_extract() {
-    // Set alignment relative to "align_to_ref" for all related memory operations.
-    for (int i = memops.size() - 1; i >= 0; i--) {
-      MemNode* s = memops.at(i)->as_Mem();
-      if (isomorphic(s, mem_ref) &&
-           (!_do_vector_loop || same_origin_idx(s, mem_ref))) {
-        const VPointer& p2 = vpointer(s);
-        if (p2.comparable(align_to_ref_p)) {
-          int align = memory_alignment(s, iv_adjustment);
-          set_alignment(s, align);
-        }
-      }
-    }
-
-    // Create initial pack pairs of memory operations for which alignment was set.
-    for (uint i = 0; i < memops.size(); i++) {
-      Node* s1 = memops.at(i);
-      int align = alignment(s1);
-      if (align == top_align) continue;
-      for (uint j = 0; j < memops.size(); j++) {
-        Node* s2 = memops.at(j);
-        if (alignment(s2) == top_align) continue;
-        if (s1 != s2 && are_adjacent_refs(s1, s2)) {
-          if (stmts_can_pack(s1, s2, align)) {
-            if (!_do_vector_loop || same_origin_idx(s1, s2)) {
-              _pairset.add_pair(s1, s2);
-            }
-          }
-        }
-      }
+  // For each group, find the adjacent memops:
+  int group_start = 0;
+  while (group_start < vpointers.length()) {
+    int group_end = group_start + 1;
+    while (group_end < vpointers.length() &&
+           VPointer::cmp_for_sort_by_group(
+             vpointers.adr_at(group_start),
+             vpointers.adr_at(group_end)
+           ) == 0) {
+      group_end++;
     }
-
-    // Remove used mem nodes.
-    for (int i = memops.size() - 1; i >= 0; i--) {
-      MemNode* m = memops.at(i)->as_Mem();
-      if (alignment(m) != top_align) {
-        memops.remove(i);
-      }
-    }
-  } // while (memops.size() != 0)
+    find_adjacent_memop_pairs_in_group(vpointers, group_start, group_end);
+    group_start = group_end;
+  }
 
 #ifndef PRODUCT
   if (is_trace_superword_packset()) {
-    tty->print_cr("\nAfter Superword::find_adjacent_refs");
+    tty->print_cr("\nAfter Superword::find_adjacent_memop_pairs");
     _pairset.print();
   }
 #endif
 }
 
-//------------------------------find_align_to_ref---------------------------
-// Find a memory reference to align the loop induction variable to.
-// Looks first at stores then at loads, looking for a memory reference
-// with the largest number of references similar to it.
-MemNode* SuperWord::find_align_to_ref(Node_List &memops, int &idx) {
-  GrowableArray<int> cmp_ct(arena(), memops.size(), memops.size(), 0);
-
-  // Count number of comparable memory ops
-  for (uint i = 0; i < memops.size(); i++) {
-    MemNode* s1 = memops.at(i)->as_Mem();
-    const VPointer& p1 = vpointer(s1);
-    for (uint j = i+1; j < memops.size(); j++) {
-      MemNode* s2 = memops.at(j)->as_Mem();
-      if (isomorphic(s1, s2)) {
-        const VPointer& p2 = vpointer(s2);
-        if (p1.comparable(p2)) {
-          (*cmp_ct.adr_at(i))++;
-          (*cmp_ct.adr_at(j))++;
-        }
-      }
-    }
-  }
```

Comment:
Note: this used to count how many "comparable" VPointers we have for each memop. Goal: find memop with the most "comparable" VPointers, in the hope that it is the longest vector.